File: GO/Ranas.go
package main

import "fmt"

type Ranas struct {
	BaseAnfibios
}

// Constructor con parametros
func NewRanas(nombre, piel string) *Ranas {
	return &Ranas{
		BaseAnfibios: *NewBaseAnfibios(nombre, piel),
	}
}

// Constructor sin parametros
func NewRanasSinParemetros() *Ranas {
	return &Ranas{
		BaseAnfibios: *NewBaseAnfibios("Renacuajo paseador", "Delgada, frágil y permeable"),
	}
}

// Constructor copia
func NewRanasCopia(copia *Ranas) *Ranas {
	return &Ranas{
		BaseAnfibios: *NewBaseAnfibios(
			copia.Nombre,
			copia.Piel,
		),
	}
}

// Destructor
func (c *Ranas) Close() error {
	fmt.Printf("Destructor...")
	return nil
}

// Metodo de la clase padre
func (c *Ranas) Metamorfosis() string {
	return "Proceso de transformacion en cuatro etapas principales: huevo, renacuajo, renacuajo con patas y rana adulta."
}

// Metodos de las interfaces
func (c *Ranas) Habitat() string {
	return "DLugares frescos y húmedos."
}

func (c *Ranas) Reproducirse() string {
	return "Se reproducen sexualmente."
}

func (c *Ranas) Presa() string {
	return "Son presas de las aves y serpientes."
}

File: JAVASCRIPT/clases/Ranas.js
import { Anfibios } from "../clases-abstractas/Anfibios.js"
import { IHabitat } from "../interfaces/IHabitat.js"
import { IReproducirse } from "../interfaces/IReproducirse.js"
import { IPresa } from "../interfaces/IPresa.js"

export class Ranas extends Anfibios {
    // Constructor con parametros
    constructor(nombre = "Renacuajo paseador", piel = "Delgada, frágil y permeable") {
        super(nombre, piel)
    }

    // Constructor copia - Una manera de realizarlo en javascript es mediante un metodo clone
    clone() {
        return new Ranas(this.nombre, this.piel)
    }

    // Simulacion del destructor
    destroy() {
        console.log("Destructor...")
        this.nombre = false
        this.piel = false
    }

    // Metodo de la clase padre
    metamorfosis() {
        console.log("Proceso de transformacion en cuatro etapas principales: huevo, renacuajo, renacuajo con patas y rana adulta.")
    }

    // Metodos de las interfaces
    habitat() {
        console.log("Lugares frescos y húmedos.")
    }

    reproducirse() {
        console.log("Se reproducen sexualmente.")
    }

    presa() {
        console.log("Son presas de las aves y serpientes.")
    }
}